File: codepropertygraph/src/main/java/io/shiftleft/codepropertygraph/cpgloading/ProtoToOverflowDb.java
package io.shiftleft.codepropertygraph.cpgloading;

import gnu.trove.set.TLongSet;
import gnu.trove.set.hash.TLongHashSet;
import io.shiftleft.proto.cpg.Cpg;
import org.apache.tinkerpop.gremlin.tinkergraph.storage.OndiskOverflow;

import java.io.File;
import java.io.IOException;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Spliterator;
import java.util.Spliterators;
import java.util.stream.StreamSupport;

public class ProtoToOverflowDb extends ProtoToX<File> {
  private final OndiskOverflow overflowDb;
  private final Map<Long, Map<String, TLongSet>> inEdgesByNodeId = new HashMap<>(); //NodeId -> EdgeLabel -> EdgeIds
  private final Map<Long, Map<String, TLongSet>> outEdgesByNodeId = new HashMap<>(); //NodeId -> EdgeLabel -> EdgeIds
  private final ProtoNodeSerializer nodeSerializer = new ProtoNodeSerializer(inEdgesByNodeId, outEdgesByNodeId);
  private final ProtoEdgeSerializer edgeSerializer = new ProtoEdgeSerializer();
  private Collection<Cpg.CpgStruct.Node> nodes;

  public ProtoToOverflowDb() throws IOException {
    this(File.createTempFile("overflowdb", "bin"));
  }

  public ProtoToOverflowDb(File writeTo) {
    this.overflowDb = OndiskOverflow.createWithSpecificLocation(writeTo);
  }

  @Override
  public void addNodes(Collection<Cpg.CpgStruct.Node> nodes) {
    /** proto nodes don't have their adjacent edges, but that's required for the OverflowDb serializer,
     * so we can only bring it together in the final `build` step */
    this.nodes = nodes;
  }

  @Override
  public void addEdges(Collection<Cpg.CpgStruct.Edge> edges) {
    final int characteristics = Spliterator.IMMUTABLE | Spliterator.CONCURRENT | Spliterator.NONNULL;
    final boolean parallel = true;
    final Spliterator<Cpg.CpgStruct.Edge> spliterator = Spliterators.spliterator(edges, characteristics);
    StreamSupport.stream(spliterator, parallel).forEach(edge -> {
      final ProtoEdgeWithId edgeWithId = new ProtoEdgeWithId(edge);
      final String label = edge.getType().name();

      // populate maps that are shared with `nodeSerializer`
      inEdgesByNodeId
          .computeIfAbsent(edgeWithId.edge.getSrc(), z -> new HashMap<>())
          .computeIfAbsent(label, z -> new TLongHashSet())
          .add(edgeWithId.id);
      outEdgesByNodeId
          .computeIfAbsent(edgeWithId.edge.getDst(), z -> new HashMap<>())
          .computeIfAbsent(label, z -> new TLongHashSet())
          .add(edgeWithId.id);

      try {
        final byte[] bytes = edgeSerializer.serialize(edgeWithId);
        overflowDb.getEdgeMVMap().put(edgeWithId.id, bytes);
      } catch (IOException e) {
        logger.error("error while serializing " + edge, e);
      }
    });
  }

  @Override
  public File build() {
    // now that we have the edges, we can finally serialize the nodes
    assert nodes != null;
    final int characteristics = Spliterator.IMMUTABLE | Spliterator.CONCURRENT | Spliterator.NONNULL;
    final boolean parallel = true;
    final Spliterator<Cpg.CpgStruct.Node> spliterator = Spliterators.spliterator(nodes, characteristics);
    StreamSupport.stream(spliterator, parallel).forEach(node -> {
      try {
        final byte[] bytes = nodeSerializer.serialize(node);
        overflowDb.getVertexMVMap().put(node.getKey(), bytes);
      } catch (IOException e) {
        logger.error("error while serializing " + node, e);
      }
    });

    overflowDb.close();
    return overflowDb.getStorageFile();
  }

}

File: codepropertygraph/src/main/scala/io/shiftleft/codepropertygraph/cpgloading/ProtoToOverflowDb.scala
package io.shiftleft.codepropertygraph.cpgloading

import java.io.{File, FileInputStream, IOException}
import java.nio.file.Files
import java.util.{HashMap => JHashMap, Map => JMap}

import gnu.trove.set.TLongSet
import gnu.trove.set.hash.TLongHashSet
import io.shiftleft.proto.cpg.Cpg.CpgStruct
import org.apache.logging.log4j.LogManager
import org.apache.tinkerpop.gremlin.tinkergraph.storage.OndiskOverflow
import scala.collection.JavaConverters._
import resource.managed

/**
  * converts cpg.bin.zip proto to OverflowDb, so we don't need to first import into Tinkergraph and then serialize it out again
  * Useful especially for large cpgs. Uses scala parallel collections.
  *
  * Usage example: create the overflowdb and load it in ocular. This will load all references in memory, and accessed elements/properties will be lazily fetched from disk.
  * ./proto2overflowdb.sh --cpg cpg.bin.zip --out overflowdb.bin
  * ./ocular.sh
  * Cpg.withStorage("overflowdb.bin")
  */
object ProtoToOverflowDb extends App {
  type NodeId = java.lang.Long
  type EdgeLabel = String

  val logger = LogManager.getLogger(getClass)
  val edgeSerializer = new ProtoEdgeSerializer

  parseConfig.map { config =>
    val writeTo = config.writeTo.getOrElse(new File("overflowdb.bin"))
    logger.info(s"running ProtoToOverflowDb with cpg=${config.cpg}; writing results to $writeTo")
    if (writeTo.exists) writeTo.delete()
    val start = System.currentTimeMillis

    val tempDir = Files.createTempDirectory("cpg2sp_proto").toFile
    try {
      ProtoCpgLoader.extractIntoTemporaryDirectory(config.cpg.getAbsolutePath, tempDir.getAbsolutePath)
      for (overflowDb <- managed(OndiskOverflow.createWithSpecificLocation(writeTo))) {
        tempDir.listFiles.filter(_.isFile).par.foreach(importProtoBin(overflowDb))
      }
      logger.info("OverflowDb construction finished in " + (System.currentTimeMillis - start) + "ms.")
    } finally ProtoCpgLoader.removeTemporaryDirectory(tempDir)
  }

  def importProtoBin(overflowDb: OndiskOverflow)(protoFile: File): Unit =
    for (inputStream <- managed(new FileInputStream(protoFile))) {
      importCpgStruct(CpgStruct.parseFrom(inputStream), overflowDb)
    }

  def importCpgStruct(cpgProto: CpgStruct, overflowDb: OndiskOverflow): Unit = {
    /** cpg proto nodes don't know their adjacent edges, but those are required for the OverflowDb serializer,
      * so we need to build some helper maps to import the nodes */
    val inEdgesByNodeId: JMap[NodeId, JMap[EdgeLabel, TLongSet]] = new JHashMap
    val outEdgesByNodeId: JMap[NodeId, JMap[EdgeLabel, TLongSet]] = new JHashMap

    cpgProto.getEdgeList.asScala.par.foreach { edge =>
      val edgeWithId = new ProtoEdgeWithId(edge)
      val label = edge.getType.name
      inEdgesByNodeId
        .computeIfAbsent(edgeWithId.edge.getSrc, _ => new JHashMap)
        .computeIfAbsent(label, _ => new TLongHashSet)
        .add(edgeWithId.id)
      outEdgesByNodeId
        .computeIfAbsent(edgeWithId.edge.getDst, _ => new JHashMap)
        .computeIfAbsent(label, _ => new TLongHashSet)
        .add(edgeWithId.id)
      overflowDb.getEdgeMVMap.put(edgeWithId.id, edgeSerializer.serialize(edgeWithId))
    }

    val nodeSerializer = new ProtoNodeSerializer(inEdgesByNodeId, outEdgesByNodeId)
    cpgProto.getNodeList.asScala.par.foreach { node =>
      overflowDb.getEdgeMVMap.put(node.getKey, nodeSerializer.serialize(node))
    }
  }

  def parseConfig: Option[Config] = {
    new scopt.OptionParser[Config](getClass.getSimpleName) {
      opt[File]("cpg").required
        .action((x, c) => c.copy(cpg = x))
        .text("path to cpg.bin.zip")
      opt[File]("out")
        .action((x, c) => c.copy(writeTo = Some(x)))
        .text("where to write overflowdb.bin output to (will delete if exists)")
      help("help").text("prints this usage text")
    }.parse(args, Config(cpg = null))
  }

}

case class Config(cpg: File, writeTo: Option[File] = None)
